### PR TITLE
Allow unitPlacementOnlyAllowedIn to be changed via a trigger in the XML

### LINKED ISSUE #14863: TriggerAttachment#getPropertyChangeHistoryStartEvent:450 - java.lang.UnsupportedOperationException
## User Description
Tried to change the value "unitPlacementOnlyAllowedIn" from a unit attachment in the XML using a trigger. The error comes up right at the moment when the trigger is called. Other unit attachment properties can be changed this way just fine, apparently not "unitPlacementOnlyAllowedIn"

## Map
north_africa / North Africa - Rommel's Last Push

## Log Message
triggerAttachmentSubmarineIndianOcean: Unable to get the value for unitPlacementOnlyAllowedIn from unitAttachment

## TripleA Version
[2026-08.16.15727](https://github.com/triplea-game/triplea/releases/tag/2026-08.16.15727)

## Java Version
25.0.3

## Operating System
Windows 11

## Heap Size
7980M

## Stack Trace
```
Exception: java.lang.UnsupportedOperationException triggerAttachmentSubmarineIndianOcean: Unable to get the value for unitPlacementOnlyAllowedIn from unitAttachment
java.lang.Exception
	at games.strategy.triplea.attachments.TriggerAttachment.getPropertyChangeHistoryStartEvent(TriggerAttachment.java:450)
	at games.strategy.triplea.attachments.TriggerAttachment.triggerUnitPropertyChange(TriggerAttachment.java:846)
	at games.strategy.triplea.attachments.TriggerAttachment.fireTriggers(TriggerAttachment.java:298)
	at games.strategy.triplea.attachments.TriggerAttachment.collectAndFireTriggers(TriggerAttachment.java:224)
	at games.strategy.triplea.delegate.BaseTripleADelegate.triggerWhenTriggerAttachments(BaseTripleADelegate.java:81)
	at games.strategy.triplea.delegate.BaseTripleADelegate.start(BaseTripleADelegate.java:32)
	at games.strategy.engine.framework.ServerGame.startStep(ServerGame.java:599)
	at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:459)
	at games.strategy.engine.framework.ServerGame.runNextStep(ServerGame.java:352)
	at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:319)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.launchInternal(LocalLauncher.java:91)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.lambda$launch$0(LocalLauncher.java:59)
	at java.base/java.lang.Thread.run(Thread.java:1474)


Exception: java.lang.UnsupportedOperationException No Getter has been defined!
java.lang.Exception
	at games.strategy.engine.data.MutableProperty.lambda$noGetter$0(MutableProperty.java:46)
	at games.strategy.engine.data.MutableProperty.getValue(MutableProperty.java:103)
	at java.base/java.util.Optional.map(Optional.java:260)
	at games.strategy.engine.data.DefaultAttachment.getRawPropertyString(DefaultAttachment.java:132)
	at games.strategy.triplea.attachments.TriggerAttachment.getPropertyChangeHistoryStartEvent(TriggerAttachment.java:447)
	at games.strategy.triplea.attachments.TriggerAttachment.triggerUnitPropertyChange(TriggerAttachment.java:846)
	at games.strategy.triplea.attachments.TriggerAttachment.fireTriggers(TriggerAttachment.java:298)
	at games.strategy.triplea.attachments.TriggerAttachment.collectAndFireTriggers(TriggerAttachment.java:224)
	at games.strategy.triplea.delegate.BaseTripleADelegate.triggerWhenTriggerAttachments(BaseTripleADelegate.java:81)
	at games.strategy.triplea.delegate.BaseTripleADelegate.start(BaseTripleADelegate.java:32)
	at games.strategy.engine.framework.ServerGame.startStep(ServerGame.java:599)
	at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:459)
	at games.strategy.engine.framework.ServerGame.runNextStep(ServerGame.java:352)
	at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:319)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.launchInternal(LocalLauncher.java:91)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.lambda$launch$0(LocalLauncher.java:59)
	at java.base/java.lang.Thread.run(Thread.java:1474)


```

### PR DESCRIPTION
**Pull request by VictoryFirst2**
## Notes to Reviewer
Initially, only the unit attachment "unitPlacementRestrictions" could be modified via a trigger in the XML. This PR allows you to modify "unitPlacementOnlyAllowedIn" as well. In the original code, it was stated that a getter for unitPlacementOnlyAllowedIn was purposely left out as it was the inverse of unitPlacementRestrictions. However, when working on a TripleA map "North Africa - Rommel's Last Push", I encountered a situation where you want to force a player to only be able to place a unit in a specific territory. If you need to use unitPlacementRestrictions for this, you'd need to list all the territories of the entire game except for that territory. It'd be much easier if you could use unitPlacementOnlyAllowedIn and state the territory you can only place into. This territory needs to be changed via triggers a couple of times to follow the game rules. This PR allows you to do that with unitPlacementOnlyAllowedIn. In short, the following XML code:

```
	<attachment name="triggerAttachmentDestroyerAtlanticOcean" attachTo="British" javaClass="TriggerAttachment" type="player">
		<option name="conditions" value="conditionAttachmentAlwaysTrue"/>
		<option name="unitAttachmentName" value="UnitAttachment" count="unitAttachment"/>
		<option name="unitType" value="UK-destroyer"/>
		<option name="unitProperty" value="unitPlacementOnlyAllowedIn" count="Atlantic Ocean"/>
		<option name="when" value="before:britishAtlanticPlace"/>
	</attachment>
```
would generate an error message, but now it does not anymore and does exactly what it is supposed to do.

Attempts to fix: https://github.com/triplea-game/triplea/issues/14863

There might be better ways to implement this. I used the help of a LLM, and the code it came up with might not be ideal. It does work as intended though. I'd be happy to discuss how to improve this.

## Testing
Added five unit tests:
1) Checks if setting unitPlacementOnlyAllowedIn actually sets said property, and also correctly updates unitPlacementRestrictions
2) Checks if setting unitPlacementRestrictions updates unitPlacementOnlyAllowedIn
3) Checks if setting both at the same time results in the last-set property determining the final value.
4) Checks if resetting unitPlacementOnlyAllowedIn results in clearing unitPlacementRestrictions and filling unitPlacementOnlyAllowedIn with all territories on the map
5) Checks if resetting unitPlacementRestrictions results in clearing it and filling unitPlacementsOnlyAlllowedIn with all territories on the map

### CODE REVIEW

**Comment by VictoryFirst2** [comment]
This is a review from Copilot:

---
Merge readiness & risk assessment
Readiness: Mostly ready; code is small and localized.

Risks:
Export format change: allUnitStatsForExporter() now includes "unitPlacementOnlyAllowedIn". If other tools parse that exact output, this is a behavioral/format change—verify downstream consumers (exporters/importers).
No unit tests: missing automated verification for parsing the new property, resetting it, and the interplay with unitPlacementRestrictions.
API/behavior subtlety: the getter computes the inverse on-the-fly. That’s fine but worth documenting expectations when both properties are used (e.g., what if both are set separately via different triggers? The current storage is unitPlacementRestrictions only, so setting onlyAllowedIn is translated into restrictions; setting restrictions directly remains authoritative).

Blocking items (recommend addressing before merge):
Add unit tests that exercise:
setting unitPlacementOnlyAllowedIn via the string and the array path,
resetting unitPlacementOnlyAllowedIn,
interaction with unitPlacementRestrictions (e.g., set restrictions then getOnlyAllowedIn, and vice versa).
Confirm CI/build passes (compile-time safety: imports for Arrays/Set/HashSet should exist — ensure compilation).
Confirm no downstream exporter/consumer will break by the added exporter field (or document/change format accordingly).

Possible improvements / suggested code changes
Add unit tests (high priority)
Test cases for parsing the property from XML triggers and for round-trip behavior (setting onlyAllowedIn should set unitPlacementRestrictions to the inverse).
Clarify behavior and edge cases in comments/Javadoc
Document that unitPlacementOnlyAllowedIn is stored internally by modifying unitPlacementRestrictions (i.e., a setter for "onlyAllowedIn" creates the restriction list). Mention behavior if both properties are set by different triggers (last writer wins).
Consider method naming / overloads for parity and clarity
The new setUnitPlacementOnlyAllowedInArray name is fine, but consider overloading setUnitPlacementOnlyAllowedIn(String...) or providing a convenience varargs method to match style used elsewhere. If reflection-based lookup depends on exact names, keep this change but add a short comment explaining why both method names exist.

Minor style/consistency
In getUnitPlacementOnlyAllowedIn you create an immutable Set.of() when unitPlacementRestrictions is null; elsewhere code uses new HashSet<>(). This is harmless but slightly inconsistent — not necessary to change.
Concrete actionable review comments (for author)

Please add unit tests that cover string-array setters, reset, and the inverse relationship.
Verify CI builds and provide the test/CI output in the PR.
Add a small Javadoc comment above getUnitPlacementOnlyAllowedIn / setUnitPlacementOnlyAllowedInArray explaining the internal representation (we store restrictions; onlyAllowedIn is translated to restrictions).
Confirm the export change (allUnitStatsForExporter) is acceptable. If this alters an external format that other tools rely on, document that or gate it behind a versioned change.

---

Implemented the following suggestions:
- Added the unit tests
- Used overloading for setUnitPlacementOnlyAllowedInArray -> setUnitPlacementOnlyAllowedIn(String...)

**Comment by DanVanAtta** [comment]
@VictoryFirst2 could you check the two issues Claude thinks are present:

Claude wrote:   

Overall verdict: Real fix (property was write-only, triggers threw); two issues to settle before merge — an exporter format change and a test-vs-production-path gap.

Makes the derived property genuinely round-trippable (getter/reset/typed-setter), fixing the original trigger error. Good behavioral coverage for the array path.

Two blockers

- F-01 (med): the exporter now unconditionally prints unitPlacementOnlyAllowedIn: for every unit on every map — and for unrestricted units (the common case) that's the entire territory list. Noisy, and it changes a fixed-format export string (Copilot flagged this too). Fix: only emit when unitPlacementRestrictions != null, or drop the line (it's derivable). Confirm no import/export consumer depends on the field set.
- F-02 (med): all 5 tests drive the String[] setter, but the real trigger path goes through the String (colon-delimited) setter — so the actual production path and the exporter output are untested. Add a colon-string-path test.

**Comment by VictoryFirst2** [comment]
@DanVanAtta Thanks. It's interesting because I wanted to test if Claude is correct about 1), but while doing that I found out that the exporter is bugged. It does not generate stats past the winners because of "could not find EndRound[]" (or something along those lines). So all the production and unit stats are missing from the file. 

In any case, once removed this part from the exporting code, it did show a list of all the territories for unitPlacementOnlyAllowedIn when unitPlacementRestrictions is empty. I think it's just the easiest to drop it as it's derivable and no one will look at it anyway.

Will add a unit test for the colon-delimited setter, and also see if I can create a bug report for the above... potentially fix it right away if it's easy.

EDIT Please see new commit.

File: game-app/game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java [hunk 1]
```diff
@@ -297,9 +297,22 @@ public class UnitAttachment extends DefaultAttachment {
...
   // a colon delimited list of territories where this unit may not be placed
-  // also an allowed setter is "setUnitPlacementOnlyAllowedIn",
-  // which just creates unitPlacementRestrictions with an inverted list of territories
   @Getter private @Nullable String[] unitPlacementRestrictions = null;
+
+  // a colon delimited list of territories where this unit may only be placed. Essentially the
+  // inverse of unitPlacementRestrictions
+  private String[] getUnitPlacementOnlyAllowedIn() {
+    final Set<String> restrictedTerritories =
+        unitPlacementRestrictions == null
```
[the remark this answers is not in the document]
Comment (VictoryFirst2) [reply game-app/game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java:306]: Implemented this

File: game-app/game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java [hunk 2]
```diff
@@ -297,9 +297,22 @@ public class UnitAttachment extends DefaultAttachment {
   // this unit to move into a territory. (units must be owned by player, not be disabled)
   private @Nullable List<String[]> requiresUnitsToMove = null;
   // a colon delimited list of territories where this unit may not be placed
-  // also an allowed setter is "setUnitPlacementOnlyAllowedIn",
-  // which just creates unitPlacementRestrictions with an inverted list of territories
   @Getter private @Nullable String[] unitPlacementRestrictions = null;
+
+  // a colon delimited list of territories where this unit may only be placed. Essentially the
+  // inverse of unitPlacementRestrictions
+  private String[] getUnitPlacementOnlyAllowedIn() {
```
[the remark this answers is not in the document]
Comment (VictoryFirst2) [reply game-app/game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java:304]: This is beyond the scope of this PR, but it nice to have this flagged.
